packages/stream/README.md

```diff
@@ -73,6 +73,33 @@ which promises settle.
 A stream is consequently a pair of queues that transport iteration results,
 one to send messages forward and another to receive acknowledgements.
 
+An async queue is itself an input and output pair.
+The `put` function makes it an async sink, and the `get` function makes it an
+async spring.
+
+## Topic
+
+The `makeTopic` function returns a kit with a `publisher` and a `subscribe`
+function.
+The `publisher` is a writer, but provides no back-pressure.
+The `subscribe` function returns readers.
+
+```js
+const { publisher, subscribe } = makeTopic();
+const subscriber = subscribe();
+publisher.next(value);
+for (const value of subscriber) {}
+```
+
+Topics are very similar to pipes, but instead of using `makeQueue`,
+topics use a very similar `makePubSub()`, which produces one sink and any
+number of springs encapsulating a shared async linked list.
+The topic writer is a stream constructed from the sink and a null spring.
+The null spring provides no forward-pressure.
+The topic readers are streams constructed from a null sink (so no
+back-pressure) and a subscriber spring that serves as an independent
+cursor starting with the next published value.
```

Comment:
To avoid unnecessary incompatibility, subscribers should start iterating at the head of the async list (i.e. a promise for the latest published value, which is only the next published one if the async list is currently empty)

Reply:
What do you think of:

- `makeChangeTopic()` subscribers start with the next change.
- `makeUpdateTopic()` subscribers start with the last published value, or wait for the first published value.
- `makeLatestTopic()` subscribers start with the last published value and may miss intermediate updates.

`makeChangeTopic()` is useful for Endo’s inbox, which looks like:

```js
const requests = new Map();
const requestsTopic = makeTopic();
async function* followInbox() {
  const pendingRequests = Array.from(requests.values());
  const subsequentRequests = requestsTopic.subscribe();
  yield* pendingRequests;
  yield* subsequentRequests();
}
```

It doesn’t fit the `makeUpdateTopic` or `makeLatestTopic` idiom because the range is discrete changes instead of replacement values.

Reply:
I'm thinking that the style of iteration should be determined entirely by the subscriber (implementation and names can be bikeshodden):

* `subscribeAllFuture(topic)` - get all future published values
* `subscribeSomeFuture(topic)` - get some (i.e. skip when possible) future published values
* `subscribeLatestThenAllFuture(topic)` - get the latest published value followed by all future published values
* `subscribeLatestThenSomeFuture(topic)` - get the latest published value, followed by some (skip when possible) future published values

As for the topics, I believe there are only push and pull modes needed (again, the implementation should be idiomatic to your other abstractions in `@endo/stream`):

* `const { pusher, topic } = makePushTopicKit();` - a topic where the publisher calls `pusher.next(valueP)` to append the corresponding value to `topic`'s async list
* `const topic = makePullTopic(puller);` - a topic which calls `puller.next()` to get a new value to append to the async list on demand (when the first subscriber reads past its end)

Reply:
I disagree. I believe the creator of the topic stands in a position where they know the semantics of the value they are publishing and with those semantics there is a natural subscription pattern. For example, if I were to publish ETAs or snapshots, I should use `makeLatestTopic`. If I were to publish deltas, patches, or operational transforms, I should use `makeChangeTopic`, where some intermediary is in a position to synchronously grab a snapshot and a change subscription in the same term. Agoric notifiers are a third category, with behavior like I propose for `makeUpdateTopic`, which is suitable for updating snapshots but doesn’t shed load like `makeLatestTopic` would.